File: backend/utils/Generator.go
package utils

import (
	"strconv"

	"golang.org/x/crypto/bcrypt"
)

type IGenerator interface {
	GenerateToken(idUser int64, password string) (string, error)
	HashPassword(password string) (string, error)
	CheckPassword(hashedPassword, password string) error
}

func NewGenerator() IGenerator {
	return &generator{}
}

type generator struct {
}

func (c generator) GenerateToken(idUser int64, password string) (string, error) {
	bytes_ := []byte(strconv.FormatInt(idUser, 10) + password)
	hash, err := bcrypt.GenerateFromPassword(bytes_, bcrypt.DefaultCost)
	return string(hash), err
}

func (c generator) HashPassword(password string) (string, error) {
	bytes_ := []byte(password)
	hash, err := bcrypt.GenerateFromPassword(bytes_, bcrypt.DefaultCost)
	return string(hash), err
}

func (c generator) CheckPassword(hashedPassword, password string) error {
	return bcrypt.CompareHashAndPassword([]byte(hashedPassword), []byte(password))
}

File: backend/src/utils/Generator.ts
import bcrypt from 'bcrypt'

export interface IGenerator {
    generateToken(idUser: number): Promise<string>
    hashPassword(password: string): Promise<string>
    comparePassword(password: string, passwordHash: string): Promise<boolean>
    generateRandomString(length: number): Promise<string>
}

export class Generator implements IGenerator {
    public async generateToken(idUser: number) {
        return bcrypt.hash(idUser.toString(), 10)
    }

    public async hashPassword(password: string) {
        return bcrypt.hash(password, 10)
    }

    public async comparePassword(password: string, passwordHash: string) {
        return bcrypt.compare(password, passwordHash)
    }

    public async generateRandomString(length: number) {
        let result = '';
        const characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
        const charactersLength = characters.length;
        let counter = 0;
        while (counter < length) {
            result += characters.charAt(Math.floor(Math.random() * charactersLength));
            counter += 1;
        }
        return result;
    }
}


